R/SimLongitudinalGSF.R

```diff
@@ -0,0 +1,188 @@
+        }
+        if (length(object@sigma) != 1) {
+            return("The parameter `sigma` must have length 1.")
+        }
+        if (length(c(object@omega_b, object@omega_s, object@omega_g)) != 3) {
+            return("The parameters `omega_b`, `omega_s`, and `omega_g` must be length 1.")
+        }
+        if (length(c(object@link_dsld, object@link_ttg, object@link_identity)) != 3) {
+            return("The parameters `link_dsld`, `link_ttg`, and `link_identity` must be length 1.")
+        }
+        return(TRUE)
+    }
+)
+
+
+#' @rdname sampleObservations
+#' @export
+sampleObservations.SimLongitudinalGSF <- function(object, times_df) {
+    times_df |>
+        dplyr::mutate(mu_sld = gsf_sld(.data$time, .data$psi_b, .data$psi_s, .data$psi_g, .data$psi_phi)) |>
+        dplyr::mutate(dsld = gsf_dsld(.data$time, .data$psi_b, .data$psi_s, .data$psi_g, .data$psi_phi)) |>
+        dplyr::mutate(ttg = gsf_ttg(.data$time, .data$psi_b, .data$psi_s, .data$psi_g, .data$psi_phi)) |>
+        dplyr::mutate(sld = stats::rnorm(dplyr::n(), .data$mu_sld, .data$mu_sld * object@sigma)) |>
+        dplyr::mutate(
+            log_haz_link =
+                (object@link_dsld * .data$dsld) +
+                (object@link_ttg * .data$ttg) +
+                (object@link_identity * .data$mu_sld)
+        )
+}
+
+
+#' @rdname sampleSubjects
+#' @export
+sampleSubjects.SimLongitudinalGSF <- function(object, subjects_df) {
+    assert_that(
+        is.factor(subjects_df$study),
+        is.factor(subjects_df$arm),
+        length(levels(subjects_df$study)) == length(object@mu_b),
+        length(levels(subjects_df$arm)) == length(object@mu_s)
+    )
+
+    res <- subjects_df |>
+        dplyr::distinct(.data$pt, .data$arm, .data$study) |>
+        dplyr::mutate(study_idx = as.numeric(.data$study)) |>
+        dplyr::mutate(arm_idx = as.numeric(.data$arm)) |>
+        dplyr::mutate(psi_b = stats::rlnorm(dplyr::n(), log(object@mu_b[.data$study_idx]), object@omega_b)) |>
+        dplyr::mutate(psi_s = stats::rlnorm(dplyr::n(), log(object@mu_s[.data$arm_idx]), object@omega_s)) |>
+        dplyr::mutate(psi_g = stats::rlnorm(dplyr::n(), log(object@mu_g[.data$arm_idx]), object@omega_g)) |>
+        dplyr::mutate(psi_phi = stats::rbeta(dplyr::n(), object@a_phi[.data$arm_idx], object@b_phi[.data$arm_idx]))
+
+    res[, c("pt", "arm", "study", "psi_b", "psi_s", "psi_g", "psi_phi")]
+}
+
+
+
+
+## sim_lm_gsf ----
+
+#' Generalized Stein-Fojo Functionals
+#'
+#' @param time (`numeric`)\cr time grid.
+#' @param b (`number`)\cr baseline.
+#' @param s (`number`)\cr shrinkage.
+#' @param g (`number`)\cr growth.
+#' @param phi (`number`)\cr shrinkage proportion.
+#'
+#' @returns The function results.
+#'
+#' @keywords internal
+gsf_sld <- function(time, b, s, g, phi) {
+    phi <- dplyr::if_else(time >= 0, phi, 0)
+    b * (phi * exp(-s * time) + (1 - phi) * exp(g * time))
+}
+
+
+#' @rdname gsf_sld
+gsf_ttg <- function(time, b, s, g, phi) {
```

Comment:
these kinds of formulas would be good to document in math in the relevant vignette (or precursor design doc) - maybe you have that already though, just thinking out loud here